Teapot/Logger.swift

```diff
@@ -0,0 +1,106 @@
+    /// NOTE: In production, you should almost certainly use `.none`, which is also the default value.
+    public var currentLevel: LogLevel
+
+    /// Designated initializer.
+    ///
+    /// - Parameter level: The level to log at. Defaults to `.none`
+    public init(level: LogLevel = .none) {
+        self.currentLevel = level
+    }
+
+    /// Logs an item if the log level is `incomingAndOutgoing` or higher.
+    ///
+    /// - Parameters:
+    ///   - items: What you wish to log.
+    ///   - file: The file this log is coming from. Defaults to the directly calling file.
+    ///   - line: The line this log is coming from. Defaults to the directly calling line.
+    ///   - date: The date this
+    /// - Returns: A boolean indicating whether the log printed or not, mostly for testing purposes.
+    @discardableResult
+    public func incomingAndOutgoingDataLog(_ items: @autoclosure () -> String, file: StaticString = #file, line: UInt = #line, at date: Date = Date()) -> Bool {
+        return log(level: .incomingAndOutgoingData, items: items, file: file, line: line, at: date)
+    }
+
+    /// Logs an item if the log level is `incomingOnly` or higher.
+    ///
+    /// - Parameters:
+    ///   - items: What you wish to log.
+    ///   - file: The file this log is coming from. Defaults to the directly calling file.
+    ///   - line: The line this log is coming from. Defaults to the directly calling line.
+    ///   - date: The date this
+    /// - Returns: A boolean indicating whether the log printed or not, mostly for testing purposes.
+    @discardableResult
+    public func incomingDataLog(_ items: @autoclosure () -> String, file: StaticString = #file, line: UInt = #line, at date: Date = Date()) -> Bool {
+        return log(level: .incomingData, items: items, file: file, line: line, at: date)
+    }
+
+    /// Logs an item if the log level is `error` or higher.
+    ///
+    /// - Parameters:
+    ///   - items: What you wish to log.
+    ///   - file: The file this log is coming from. Defaults to the directly calling file.
+    ///   - line: The line this log is coming from. Defaults to the directly calling line.
+    ///   - date: The date this
+    /// - Returns: A boolean indicating whether the log printed or not, mostly for testing purposes.
+    @discardableResult
+    public func errorLog(_ items: @autoclosure () -> String, file: StaticString = #file, line: UInt = #line, at date: Date = Date()) -> Bool {
+        return log(level: .error, items: items, file: file, line: line, at: date)
+    }
+
+    private func log(level: LogLevel, items: @autoclosure () -> String, file: StaticString, line: UInt, at date: Date) -> Bool {
+        guard level.rawValue >= currentLevel.rawValue else {
+            // We don't want to print anything at a level lower than the current log level.
+            return false
+        }
+
+        print("TEAPOT - \(file) line \(line) at \(date): \(items())")
```

Comment:
Why print, instead of `NSLog`? Print just outputs stuff to stdout, whereas `NSLog` actually logs stuff, so even if the process is not attached to Xcode, we can see logs in the Console.app.

Reply:
Mostly because I wanted to do my own formatting separate from what `NSLog` does to make it really clear where the log barf is coming from.

However, it also wouldn't be great from a security standpoint to have this be persisted in the console instead of just via `stdout` IMO.

Reply:
I don't think that is doable with `print`. 🤔 Maybe there's some way to hook into the console and create our own `NSLog`-like variant?

Reply:
Good point. 👍